### ISSUE TITLE
prescan snapshot sometimes hangs

### ISSUE THREAD

**Issue by pjb671** [sardana-org/sardana#1080]
Hello,
Apologies for opening more issues when you still need feedback on the first. At Nanomax at Max IV we have noticed scans sometimes failing in the prescan snapshot of gscan. The beamline recently added many more attributes to the snapshot (over 100 in total) and the failures started then. I can see very clearly that the scan simply hangs at line 765 here https://github.com/sardana-org/sardana/blob/release-Jul18/src/sardana/macroserver/scan/gscan.py (v = PyTango.AttributeProxy(column.source).read().value)
The strange thing is that it is not always the same attribute that causes the stop, and I do not believe there is any problem with the attribute when it happens. Any anyway, this is within try/catch so if the attribute could not be read there should be an error. I have a debug statement before and after the read line, so I see which attribute it is trying to read, but the read simply never returns.  And as I say, it can be any random attribute. I would say it happens maybe 1 in 20 times, perhaps once a week.
cheers,
Paul

**Comment by reszelaz** [comment]
To be honest this is a first time I hear about this kind of behavior.

For sure you must have checked this already but just in case... could you please somehow stress test the tango reads (from the same host where the MacroServer runs), something like the following snippet run in a loop:

```python
import tango
import shelve

path = "/home/zreszela/workspace/MacroServer/zreszela/macroserver.properties"
env = shelve.open(path)
snapshot = env["PreScanSnapshot"]
for item, label in snapshot:
    dev, attr = None, None
    try:
        dev = tango.DeviceProxy(item)
    except:
        attr = tango.AttributeProxy(item)
    if dev:
        if "Position" in dev.get_attribute_list():
            value = dev.read_attribute("Position").value
        elif "Value" in dev.get_attribute_list():
            value = dev.read_attribute("Value").value
    elif attr:
        value = attr.read().value
    print("{0}: {1}".format(item, value))
```
(please remember to change path to your environment location, if you don't change the `EnvironmentDb` property of the MacroServer this should be: `/tmp/tango/MacroServer/<inst_name>/macroserver.properties`).

**Comment by dschick** [comment]
I encountered the same problem. After around 200-400 scans Sardana got stuck during the snapshot.

I am currently running the snippet of @reszelaz 
Before that I tried a` %wum` motor for every motor in the snapshot group 1000 times without an issue.
I post the result here as soon as possible.

**Comment by mirlin** [comment]
We have also run the snippet many thousand times without issue (using the large prescan snapshot group at the beamline), but when the beamline scans it keeps failing. So we replaced the following line in sardana/macroserver/scan/gscan.py     
    v = PyTango.AttributeProxy(column.source).read().value

with
                try:
                    msg = "PJB taking prescan reading, deleting previous AttributeProxy: {}"
                    msg = msg.format(_attr_proxy)
                    self.debug(msg)
                    # _attr is undifinied at the first iteration, the error is catched.
                    del _attr_proxy
                    self.debug("PJB taking prescan reading, Attr Proxy deleted")
                    msg = "PJB taking prescan reading, deleting previous Attribute value: {}"
                    msg = msg.format(_attr_value)
                    self.debug(msg)
                    del _attr_value
                    self.debug("PJB taking prescan reading, Attr Value deleted")
                except NameError:
                    pass
                self.debug('PJB taking prescan, setting up device proxy')
                _attr_proxy = PyTango.AttributeProxy(column.source)
                self.debug('PJB taking prescan, reading attribute')
                _attr_value = _attr_proxy.read()
                self.debug('PJB taking prescan, getting value')
                v = _attr_value.value

to try and narrow it down and each time it fails, it is on the  del _attr_value  line.

Any ideas are very welcome... it fails several times a week.

**Comment by reszelaz** [comment]
Thanks for all the debugging and sorry about this problem. Can you also try to run this snippet within a simple device server? For example start the loop on a command execution and run it in a background thread?

**Comment by reszelaz** [comment]
Another idea could be to use the `tango.DeviceProxy.read_attribute` instead of the `tango.AttributeProxy.read`. Could you try this?

**Comment by mirlin** [comment]
Thank you for these suggestions. We have a long weekend from tomorrow due to the Swedish national day, but we will run more tests on Monday!

Cheers
Mirjam

On 5 Jun 2019 20:40, reszelaz <notifications@github.com> wrote:

Another idea could be to use the tango.DeviceProxy.read_attribute instead of the tango.AttributeProxy.read. Could you try this?

**Comment by dschick** [comment]
hej,

I am also off for the next two weeks. So for now we will simply skip the snapshot.

I am not 100% sure, but I have not encountered the bug with our second setup, which does not have any TANGO motors - all hardware is directly implemented in Sardana Controllers.

Best

Daniel

**Comment by daneos** [comment]
Could you test it with #1146? We had similar problem, and it seems this PR solved it for us.

**Comment by reszelaz** [comment]
> Could you test it with #1146? We had similar problem, and it seems this PR solved it for us.

Good to hear that! This would be aligned with the idea of using the `DeviceProxy` instead of `AttributeProxy` as explained in https://github.com/sardana-org/sardana/issues/1080#issuecomment-499205838. Taurus is using the `DeviceProxy.read_attribute`.

**Comment by pjb671** [comment]
I tried the taurus solution of #1080 but this seemed to cause new issues - I will reply in that thread. Then I tried a DeviceProxy instead of an AttributeProxy, and 24h later (with user operation) I saw the first crash, so it does not seem to change anything:

...many successful reads before this line, extra PJB comments added in gscan:
SardanaTP.W007 DEBUG 2019-06-20 10:34:42,543 SScan: PJB taking prescan reading: tango://b-v-nanomax-csdb-0.maxiv.lu.se:10000/B303A-EH/CTL/PZCU-04-11/Position
SardanaTP.W007 DEBUG 2019-06-20 10:34:42,543 SScan: PJB taking prescan reading, dev = tango://b-v-nanomax-csdb-0.maxiv.lu.se:10000/B303A-EH/CTL/PZCU-04-11
SardanaTP.W007 DEBUG 2019-06-20 10:34:42,543 SScan: PJB taking prescan reading, att = Position

no further output. hangs at this point. Note the log shows how I split up the device and attr, and the proxy is to the device tango://b-v-nanomax-csdb-0.maxiv.lu.se:10000/B303A-EH/CTL/PZCU-04-11. As commented before, there is no pattern in which device it fails on.

**Comment by reszelaz** [comment]
Thanks for the update @pjb671!

Have you tried to reproduce it in this way: https://github.com/sardana-org/sardana/issues/1080#issuecomment-499168269?

**Comment by reszelaz** [comment]
After tests done by @dschick in https://github.com/sardana-org/sardana/pull/1190#issuecomment-530772823 we have a strong feeling that this bug is fixed in 2.8.2 and in 3.0.1-alpha. So we are closing this bug. Please reopen if you still suffer it. Thanks to all!

**Comment by reszelaz** [comment]
We just experience it again at ALBA. There is a problem with the [`FastDummyMotorController`](https://github.com/sardana-org/sardana/blob/418b7dbe2eeb9274c5c8566e4844d6bc5212a89f/src/sardana/pool/poolcontrollers/DummyMotorController.py#L615). When you use it and make `wa` then you may be affected by this hung in the following PreScanSnapshot acquisitions.
